File: src/sqlite_db.rs
//! SQLite-backed implementation of the [`Database`] trait.

use chrono::NaiveDate;
use rusqlite::{Connection, OptionalExtension, params, types::Type};
use std::collections::BTreeMap;
use std::str::FromStr;

use crate::weight::Weight;
use crate::{
    database::{Database, DbResult},
    models::{ExecutionSet, Lift, LiftExecution, LiftRegion, LiftStats, LiftType, Muscle, SetMetric},
};

/// Current database schema version.
const DB_VERSION: i32 = 7;

/// Database persisted to a SQLite file.
pub struct SqliteDb {
    conn: Connection,
}

impl SqliteDb {
    /// Open (or create) a SQLite database at `path`.
    pub fn new(path: &str) -> DbResult<Self> {
        let conn = Connection::open(path)?;
        init_db(&conn)?;
        Ok(Self { conn })
    }

    /// Load all execution records for `lift_id`, newest first.
    fn fetch_executions(&self, lift_id: i32) -> DbResult<Vec<LiftExecution>> {
        let mut stmt = self.conn.prepare(
            "SELECT id, date, sets, notes FROM lift_records WHERE lift_id = ?1 ORDER BY date DESC",
        )?;
        let iter = stmt.query_map(params![lift_id], |row| {
            let date_str: String = row.get(1)?;
            let date = NaiveDate::parse_from_str(&date_str, "%Y-%m-%d").map_err(|e| {
                rusqlite::Error::FromSqlConversionFailure(1, Type::Text, Box::new(e))
            })?;
            let sets_json: String = row.get(2)?;
            let sets: Vec<ExecutionSet> = serde_json::from_str(&sets_json).unwrap_or_default();
            Ok(LiftExecution {
                id: Some(row.get(0)?),
                date,
                sets,
                notes: row.get(3)?,
            })
        })?;
        let mut executions = Vec::new();
        for exec in iter {
            executions.push(exec?);
        }
        Ok(executions)
    }
}

fn table_exists(conn: &Connection, table: &str) -> DbResult<bool> {
    let exists: Option<String> = conn
        .query_row(
            "SELECT name FROM sqlite_master WHERE type='table' AND name=?1",
            params![table],
            |row| row.get(0),
        )
        .optional()?;
    Ok(exists.is_some())
}

fn has_column(conn: &Connection, table: &str, column: &str) -> DbResult<bool> {
    let mut stmt = conn.prepare(&format!("PRAGMA table_info({})", table))?;
    let mut rows = stmt.query([])?;
    while let Some(row) = rows.next()? {
        let name: String = row.get(1)?;
        if name == column {
            return Ok(true);
        }
    }
    Ok(false)
}

fn detect_version(conn: &Connection) -> DbResult<i32> {
    if !table_exists(conn, "lifts")? {
        return Ok(0);
    }
    let mut v = 1;
    if has_column(conn, "lifts", "region")? {
        v = 2;
    }
    if has_column(conn, "lifts", "main_lift")? {
        v = 3;
    }
    if has_column(conn, "lifts", "muscles")? {
        v = 4;
    }
    if has_column(conn, "lifts", "notes")? {
        v = 5;
    }
    if has_column(conn, "lift_records", "notes")? {
        v = 6;
        if !has_column(conn, "lift_records", "reps")? {
            v = 7;
        }
    }
    Ok(v)
}

/// Initialize the database schema and ensure it matches the expected version.
fn init_db(conn: &Connection) -> DbResult<()> {
    let detected_version = detect_version(conn)?;
    if detected_version == 0 {
        // Fresh database with no tables yet.
        conn.execute(
            "CREATE TABLE IF NOT EXISTS lifts (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                name TEXT NOT NULL UNIQUE,
                region TEXT NOT NULL,
                main_lift TEXT,
                muscles TEXT NOT NULL,
                notes TEXT NOT NULL DEFAULT ''
            )",
            [],
        )?;
        conn.execute(
            "CREATE TABLE IF NOT EXISTS lift_records (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                lift_id INTEGER NOT NULL,
                date TEXT NOT NULL,
                sets TEXT NOT NULL,
                notes TEXT NOT NULL DEFAULT '',
                FOREIGN KEY(lift_id) REFERENCES lifts(id)
            )",
            [],
        )?;
        conn.pragma_update(None, "user_version", &DB_VERSION)?;
    } else if detected_version < DB_VERSION {
        // Upgrade legacy schemas stepwise.
        run_migrations(conn, detected_version)?;
    } else {
        // Schema matches the latest version but user_version might be incorrect.
        let user_version: i32 =
            conn.query_row("PRAGMA user_version", [], |row| row.get(0))?;
        if user_version != DB_VERSION {
            conn.pragma_update(None, "user_version", &DB_VERSION)?;
        }
    }
    Ok(())
}

/// Apply migrations from a previous schema version to [`DB_VERSION`].
fn run_migrations(conn: &Connection, mut from_version: i32) -> DbResult<()> {
    // Apply migrations sequentially so that every intermediate version is handled.
    while from_version < DB_VERSION {
        match from_version {
            1 => {
                // Version 1 stored weight as a numeric column and lacked lift metadata.
                conn.execute(
                    "ALTER TABLE lifts ADD COLUMN region TEXT NOT NULL DEFAULT 'UPPER'",
                    [],
                )?;
                conn.execute(
                    "ALTER TABLE lift_records RENAME COLUMN weight TO weight_old",
                    [],
                )?;
                conn.execute(
                    "ALTER TABLE lift_records ADD COLUMN weight TEXT NOT NULL DEFAULT '0'",
                    [],
                )?;
                conn.execute("UPDATE lift_records SET weight = weight_old", [])?;
                conn.execute("ALTER TABLE lift_records DROP COLUMN weight_old", [])?;
            }
            2 => {
                // Add optional designation for a lift's main variant.
                conn.execute("ALTER TABLE lifts ADD COLUMN main_lift TEXT", [])?;
            }
            3 => {
                // Track muscles targeted by each lift.
                conn.execute(
                    "ALTER TABLE lifts ADD COLUMN muscles TEXT NOT NULL DEFAULT ''",
                    [],
                )?;
            }
            4 => {
                // Free-form notes for lifts were introduced in version 5.
                conn.execute(
                    "ALTER TABLE lifts ADD COLUMN notes TEXT NOT NULL DEFAULT ''",
                    [],
                )?;
            }
            5 => {
                // Execution records also store notes starting with version 6.
                conn.execute(
                    "ALTER TABLE lift_records ADD COLUMN notes TEXT NOT NULL DEFAULT ''",
                    [],
                )?;
            }
            6 => {
                // Migrate execution records to store set details as JSON.
                conn.execute("ALTER TABLE lift_records RENAME TO lift_records_old", [])?;
                conn.execute(
                    "CREATE TABLE lift_records (
                        id INTEGER PRIMARY KEY AUTOINCREMENT,
                        lift_id INTEGER NOT NULL,
                        date TEXT NOT NULL,
                        sets TEXT NOT NULL,
                        notes TEXT NOT NULL DEFAULT '',
                        FOREIGN KEY(lift_id) REFERENCES lifts(id)
                    )",
                    [],
                )?;
                let mut stmt = conn.prepare(
                    "SELECT id, lift_id, date, sets, reps, weight, rpe, notes FROM lift_records_old",
                )?;
                let mut rows = stmt.query([])?;
                while let Some(row) = rows.next()? {
                    let id: i32 = row.get(0)?;
                    let lift_id: i32 = row.get(1)?;
                    let date: String = row.get(2)?;
                    let set_count: i32 = row.get(3)?;
                    let reps: i32 = row.get(4)?;
                    let weight_str: String = row.get(5)?;
                    let rpe: Option<f32> = row.get(6)?;
                    let notes: String = row.get(7)?;
                    let weight = Weight::from_str(&weight_str).unwrap_or(Weight::Raw(0.0));
                    let set = ExecutionSet { metric: SetMetric::Reps(reps), weight: weight.clone(), rpe };
                    let sets = vec![set; set_count as usize];
                    let sets_json = serde_json::to_string(&sets)?;
                    conn.execute(
                        "INSERT INTO lift_records (id, lift_id, date, sets, notes) VALUES (?1, ?2, ?3, ?4, ?5)",
                        params![id, lift_id, date, sets_json, notes],
                    )?;
                }
                conn.execute("DROP TABLE lift_records_old", [])?;
            }
            _ => {}
        }
        from_version += 1;
    }
    conn.pragma_update(None, "user_version", &DB_VERSION)?;
    Ok(())
}

impl Database for SqliteDb {
    fn add_lift(
        &self,
        name: &str,
        region: LiftRegion,
        main: Option<LiftType>,
        muscles: &[Muscle],
        notes: &str,
    ) -> DbResult<()> {
        let muscles_str = muscles
            .iter()
            .map(|m| m.to_string())
            .collect::<Vec<_>>()
            .join(",");
        self.conn.execute(
            "INSERT INTO lifts (name, region, main_lift, muscles, notes) VALUES (?1, ?2, ?3, ?4, ?5)",
            params![
                name,
                region.to_string(),
                main.map(|m| m.to_string()),
                muscles_str,
                notes
            ],
        )?;
        Ok(())
    }

    fn add_lift_execution(&self, name: &str, execution: &LiftExecution) -> DbResult<()> {
        let mut stmt = self.conn.prepare("SELECT id FROM lifts WHERE name = ?1")?;
        let lift_id: Option<i32> = stmt.query_row(params![name], |row| row.get(0)).optional()?;
        let lift_id = match lift_id {
            Some(id) => id,
            None => {
                return Err("lift not found".into());
            }
        };
        let sets_json = serde_json::to_string(&execution.sets)?;
        self.conn.execute(
            "INSERT INTO lift_records (lift_id, date, sets, notes) VALUES (?1, ?2, ?3, ?4)",
            params![
                lift_id,
                execution.date.to_string(),
                sets_json,
                execution.notes
            ],
        )?;
        Ok(())
    }

    fn update_lift(
        &self,
        current_name: &str,
        new_name: &str,
        region: LiftRegion,
        main: Option<LiftType>,
        muscles: &[Muscle],
        notes: &str,
    ) -> DbResult<()> {
        let muscles_str = muscles
            .iter()
            .map(|m| m.to_string())
            .collect::<Vec<_>>()
            .join(",");
        self.conn.execute(
            "UPDATE lifts SET name = ?1, region = ?2, main_lift = ?3, muscles = ?4, notes = ?5 WHERE name = ?6",
            params![
                new_name,
                region.to_string(),
                main.map(|m| m.to_string()),
                muscles_str,
                notes,
                current_name
            ],
        )?;
        Ok(())
    }

    fn update_lift_execution(&self, exec_id: i32, execution: &LiftExecution) -> DbResult<()> {
        let sets_json = serde_json::to_string(&execution.sets)?;
        self.conn.execute(
            "UPDATE lift_records SET date = ?1, sets = ?2, notes = ?3 WHERE id = ?4",
            params![
                execution.date.to_string(),
                sets_json,
                execution.notes,
                exec_id
            ],
        )?;
        Ok(())
    }

    fn lift_stats(&self, name: &str) -> DbResult<LiftStats> {
        let lift_id: i32 = self.conn.query_row(
            "SELECT id FROM lifts WHERE name = ?1",
            params![name],
            |row| row.get(0),
        )?;
        let last = self
            .conn
            .query_row(
                "SELECT id, date, sets, notes FROM lift_records WHERE lift_id = ?1 ORDER BY date DESC LIMIT 1",
                params![lift_id],
                |row| {
                    let date_str: String = row.get(1)?;
                    let date = NaiveDate::parse_from_str(&date_str, "%Y-%m-%d").map_err(|e| {
                        rusqlite::Error::FromSqlConversionFailure(1, Type::Text, Box::new(e))
                    })?;
                    let sets_json: String = row.get(2)?;
                    let sets: Vec<ExecutionSet> = serde_json::from_str(&sets_json).unwrap_or_default();
                    Ok(LiftExecution {
                        id: Some(row.get(0)?),
                        date,
                        sets,
                        notes: row.get(3)?,
                    })
                },
            )
            .optional()?;
        let mut stmt = self
            .conn
            .prepare("SELECT sets FROM lift_records WHERE lift_id = ?1")?;
        let iter = stmt.query_map(params![lift_id], |row| row.get::<_, String>(0))?;
        let mut best = BTreeMap::new();
        for row in iter {
            let sets_json = row?;
            let sets: Vec<ExecutionSet> = serde_json::from_str(&sets_json).unwrap_or_default();
            for set in sets {
                if let SetMetric::Reps(r) = set.metric {
                    let entry = best.entry(r).or_insert(set.weight.clone());
                    if set.weight.to_lbs() > entry.to_lbs() {
                        *entry = set.weight;
                    }
                }
            }
        }
        Ok(LiftStats {
            last,
            best_by_reps: best,
        })
    }

    fn list_lifts(&self, name: Option<&str>) -> DbResult<Vec<Lift>> {
        let mut lifts = Vec::new();
        if let Some(n) = name {
            let mut stmt = self.conn.prepare(
                "SELECT id, name, region, main_lift, muscles, notes FROM lifts WHERE name = ?1 ORDER BY name",
            )?;
            let iter = stmt.query_map(params![n], |row| {
                let id: i32 = row.get(0)?;
                let region_str: String = row.get(2)?;
                let region = LiftRegion::from_str(&region_str).unwrap_or(LiftRegion::UPPER);
                let main_str: Option<String> = row.get(3)?;
                let main = match main_str {
                    Some(m) => LiftType::from_str(&m).ok(),
                    None => None,
                };
                let muscles_str: String = row.get(4)?;
                let notes: String = row.get(5)?;
                let muscles = if muscles_str.is_empty() {
                    Vec::new()
                } else {
                    muscles_str
                        .split(',')
                        .filter_map(|m| Muscle::from_str(m).ok())
                        .collect()
                };
                Ok((
                    id,
                    Lift {
                        name: row.get(1)?,
                        region,
                        main,
                        muscles,
                        notes,
                        executions: Vec::new(),
                    },
                ))
            })?;
            for row in iter {
                let (id, mut lift) = row?;
                lift.executions = self.fetch_executions(id)?;
                lifts.push(lift);
            }
        } else {
            let mut stmt = self.conn.prepare(
                "SELECT id, name, region, main_lift, muscles, notes FROM lifts ORDER BY name",
            )?;
            let iter = stmt.query_map([], |row| {
                let id: i32 = row.get(0)?;
                let region_str: String = row.get(2)?;
                let region = LiftRegion::from_str(&region_str).unwrap_or(LiftRegion::UPPER);
                let main_str: Option<String> = row.get(3)?;
                let main = match main_str {
                    Some(m) => LiftType::from_str(&m).ok(),
                    None => None,
                };
                let muscles_str: String = row.get(4)?;
                let notes: String = row.get(5)?;
                let muscles = if muscles_str.is_empty() {
                    Vec::new()
                } else {
                    muscles_str
                        .split(',')
                        .filter_map(|m| Muscle::from_str(m).ok())
                        .collect()
                };
                Ok((
                    id,
                    Lift {
                        name: row.get(1)?,
                        region,
                        main,
                        muscles,
                        notes,
                        executions: Vec::new(),
                    },
                ))
            })?;
            for row in iter {
                let (id, mut lift) = row?;
                lift.executions = self.fetch_executions(id)?;
                lifts.push(lift);
            }
        }
        Ok(lifts)
    }
}

File: java/src/main/java/org/lift/trax/SqliteDb.java
package org.lift.trax;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.File;
import java.sql.*;
import java.time.LocalDate;
import java.util.*;

public class SqliteDb implements Database {
    private final Connection conn;
    private final ObjectMapper mapper = new ObjectMapper();

    public SqliteDb(String path) throws Exception {
        File dbFile = new File(path);
        this.conn = DriverManager.getConnection("jdbc:sqlite:" + path);
        if (!dbFile.exists() || dbFile.length() == 0) {
            initDb();
        }
    }

    private void initDb() throws SQLException {
        try (Statement st = conn.createStatement()) {
            st.executeUpdate("CREATE TABLE IF NOT EXISTS lifts (" +
                    "id INTEGER PRIMARY KEY AUTOINCREMENT," +
                    "name TEXT UNIQUE NOT NULL," +
                    "region TEXT NOT NULL," +
                    "main_lift TEXT," +
                    "muscles TEXT," +
                    "notes TEXT" +
                    ")");
            st.executeUpdate("CREATE TABLE IF NOT EXISTS lift_records (" +
                    "id INTEGER PRIMARY KEY AUTOINCREMENT," +
                    "lift_id INTEGER NOT NULL," +
                    "date TEXT NOT NULL," +
                    "sets TEXT NOT NULL," +
                    "warmup INTEGER NOT NULL," +
                    "notes TEXT," +
                    "FOREIGN KEY(lift_id) REFERENCES lifts(id)" +
                    ")");
        }
    }

    private List<ExecutionSet> parseSets(String json) throws Exception {
        return mapper.readValue(json, new TypeReference<List<ExecutionSet>>() {});
    }

    private String serializeSets(List<ExecutionSet> sets) throws Exception {
        return mapper.writeValueAsString(sets);
    }

    private LiftExecution mapExecution(ResultSet rs) throws Exception {
        int id = rs.getInt("id");
        LocalDate date = LocalDate.parse(rs.getString("date"));
        List<ExecutionSet> sets = parseSets(rs.getString("sets"));
        boolean warmup = rs.getInt("warmup") != 0;
        String notes = rs.getString("notes");
        return new LiftExecution(id, date, sets, warmup, notes);
    }

    private List<LiftExecution> fetchExecutions(int liftId) throws Exception {
        List<LiftExecution> list = new ArrayList<>();
        try (PreparedStatement ps = conn.prepareStatement(
                "SELECT id, date, sets, warmup, notes FROM lift_records WHERE lift_id = ? ORDER BY date DESC")) {
            ps.setInt(1, liftId);
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    list.add(mapExecution(rs));
                }
            }
        }
        return list;
    }

    private Lift rowToLift(ResultSet rs) throws Exception {
        int id = rs.getInt("id");
        String name = rs.getString("name");
        LiftRegion region = LiftRegion.valueOf(rs.getString("region"));
        String mainStr = rs.getString("main_lift");
        LiftType main = mainStr == null ? null : LiftType.fromString(mainStr);
        String musclesStr = rs.getString("muscles");
        List<Muscle> muscles = new ArrayList<>();
        if (musclesStr != null && !musclesStr.isEmpty()) {
            for (String m : musclesStr.split(",")) {
                muscles.add(Muscle.fromString(m.trim()));
            }
        }
        String notes = rs.getString("notes");
        Lift lift = new Lift(name, region, main, muscles, notes);
        lift.executions = fetchExecutions(id);
        return lift;
    }

    private List<Lift> loadLifts(String sql, Object... params) throws Exception {
        List<Lift> lifts = new ArrayList<>();
        try (PreparedStatement ps = conn.prepareStatement(sql)) {
            for (int i = 0; i < params.length; i++) {
                ps.setObject(i + 1, params[i]);
            }
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    lifts.add(rowToLift(rs));
                }
            }
        }
        return lifts;
    }

    private LiftExecution getLastExecution(int liftId) throws Exception {
        try (PreparedStatement ps = conn.prepareStatement(
                "SELECT id, date, sets, warmup, notes FROM lift_records WHERE lift_id = ? ORDER BY date DESC LIMIT 1")) {
            ps.setInt(1, liftId);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) {
                    return mapExecution(rs);
                }
            }
        }
        return null;
    }

    private Map<Integer, Double> collectBestByReps(int liftId) throws Exception {
        Map<Integer, Double> best = new TreeMap<>();
        try (PreparedStatement ps = conn.prepareStatement(
                "SELECT sets FROM lift_records WHERE lift_id = ?")) {
            ps.setInt(1, liftId);
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    List<ExecutionSet> sets = parseSets(rs.getString(1));
                    for (ExecutionSet set : sets) {
                        if (set.reps != null) {
                            int reps = set.reps;
                            double weight = set.weight;
                            best.merge(reps, weight, Math::max);
                        }
                    }
                }
            }
        }
        return best;
    }

    @Override
    public void addLift(String name, LiftRegion region, LiftType main, List<Muscle> muscles, String notes) throws Exception {
        String musclesStr = String.join(",", muscles.stream().map(Muscle::toString).toList());
        try (PreparedStatement ps = conn.prepareStatement(
                "INSERT INTO lifts (name, region, main_lift, muscles, notes) VALUES (?,?,?,?,?)")) {
            ps.setString(1, name);
            ps.setString(2, region.toString());
            ps.setString(3, main != null ? main.toString() : null);
            ps.setString(4, musclesStr);
            ps.setString(5, notes);
            ps.executeUpdate();
        }
    }

    @Override
    public void addLiftExecution(String name, LiftExecution execution) throws Exception {
        Integer liftId = null;
        try (PreparedStatement ps = conn.prepareStatement("SELECT id FROM lifts WHERE name = ?")) {
            ps.setString(1, name);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) {
                    liftId = rs.getInt(1);
                }
            }
        }
        if (liftId == null) throw new Exception("lift not found");
        String setsJson = serializeSets(execution.sets);
        try (PreparedStatement ps = conn.prepareStatement(
                "INSERT INTO lift_records (lift_id, date, sets, warmup, notes) VALUES (?,?,?,?,?)")) {
            ps.setInt(1, liftId);
            ps.setString(2, execution.date.toString());
            ps.setString(3, setsJson);
            ps.setInt(4, execution.warmup ? 1 : 0);
            ps.setString(5, execution.notes);
            ps.executeUpdate();
        }
    }

    @Override
    public void updateLift(String currentName, String newName, LiftRegion region, LiftType main, List<Muscle> muscles, String notes) throws Exception {
        String musclesStr = String.join(",", muscles.stream().map(Muscle::toString).toList());
        try (PreparedStatement ps = conn.prepareStatement(
                "UPDATE lifts SET name=?, region=?, main_lift=?, muscles=?, notes=? WHERE name=?")) {
            ps.setString(1, newName);
            ps.setString(2, region.toString());
            ps.setString(3, main != null ? main.toString() : null);
            ps.setString(4, musclesStr);
            ps.setString(5, notes);
            ps.setString(6, currentName);
            ps.executeUpdate();
        }
    }

    @Override
    public void deleteLift(String name) throws Exception {
        Integer id = null;
        try (PreparedStatement ps = conn.prepareStatement("SELECT id FROM lifts WHERE name = ?")) {
            ps.setString(1, name);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) id = rs.getInt(1);
            }
        }
        if (id != null) {
            try (PreparedStatement ps = conn.prepareStatement("DELETE FROM lift_records WHERE lift_id = ?")) {
                ps.setInt(1, id);
                ps.executeUpdate();
            }
            try (PreparedStatement ps = conn.prepareStatement("DELETE FROM lifts WHERE id = ?")) {
                ps.setInt(1, id);
                ps.executeUpdate();
            }
        }
    }

    @Override
    public void updateLiftExecution(int execId, LiftExecution execution) throws Exception {
        String setsJson = serializeSets(execution.sets);
        try (PreparedStatement ps = conn.prepareStatement(
                "UPDATE lift_records SET date=?, sets=?, warmup=?, notes=? WHERE id=?")) {
            ps.setString(1, execution.date.toString());
            ps.setString(2, setsJson);
            ps.setInt(3, execution.warmup ? 1 : 0);
            ps.setString(4, execution.notes);
            ps.setInt(5, execId);
            ps.executeUpdate();
        }
    }

    @Override
    public void deleteLiftExecution(int execId) throws Exception {
        try (PreparedStatement ps = conn.prepareStatement("DELETE FROM lift_records WHERE id = ?")) {
            ps.setInt(1, execId);
            ps.executeUpdate();
        }
    }

    @Override
    public LiftStats liftStats(String name) throws Exception {
        int liftId;
        try (PreparedStatement ps = conn.prepareStatement("SELECT id FROM lifts WHERE name = ?")) {
            ps.setString(1, name);
            try (ResultSet rs = ps.executeQuery()) {
                if (!rs.next()) throw new Exception("lift not found");
                liftId = rs.getInt(1);
            }
        }
        LiftExecution last = getLastExecution(liftId);
        Map<Integer, Double> best = collectBestByReps(liftId);
        return new LiftStats(last, best);
    }

    @Override
    public List<Lift> listLifts(String nameFilter) throws Exception {
        if (nameFilter != null) {
            return loadLifts("SELECT id, name, region, main_lift, muscles, notes FROM lifts WHERE name = ? ORDER BY name", nameFilter);
        }
        return loadLifts("SELECT id, name, region, main_lift, muscles, notes FROM lifts ORDER BY name");
    }

    @Override
    public List<Lift> liftsByType(LiftType liftType) throws Exception {
        return loadLifts("SELECT id, name, region, main_lift, muscles, notes FROM lifts WHERE main_lift = ? ORDER BY name", liftType.toString());
    }

    @Override
    public List<Lift> liftsByRegion(LiftRegion region) throws Exception {
        return loadLifts("SELECT id, name, region, main_lift, muscles, notes FROM lifts WHERE region = ? ORDER BY name", region.toString());
    }

    @Override
    public List<Lift> liftsByRegionAndType(LiftRegion region, LiftType liftType) throws Exception {
        return loadLifts("SELECT id, name, region, main_lift, muscles, notes FROM lifts WHERE region = ? AND main_lift = ? ORDER BY name", region.toString(), liftType.toString());
    }
}
